hail/python/hail/expr/functions.py

```diff
@@ -233,11 +233,17 @@ def literal(x: Any, dtype: Optional[Union[HailType, str]] = None):
     -------
     :class:`.Expression`
     """
-    wrapper = {'has_expr': False}
+    wrapper = {'has_expr': False, 'has_free_vars': False}
 
     def typecheck_expr(t, x):
         if isinstance(x, Expression):
             wrapper['has_expr'] = True
+            wrapper['has_free_vars'] = (
+                x._ir.free_vars.__len__() > 0 or
+                x._ir.free_agg_vars.__len__() > 0 or
+                x._ir.free_scan_vars.__len__() > 0
```

Comment:
This needs to be a `|=`, no?

Reply:
no difference?

Reply:
I mean `wrapper['has_free_vars'] |= ...`. Otherwise, if there are multiple nested expressions, and the last has no free variables but others do, we will only remember the `False` from the last.

Reply:
Ah yes. Mutation. I forgot this can getbcalled more than once